File: Data_Structures_And_Algorithms/C/Data_Structure/doubly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

// Doubly_Linked_List Data_Structure
typedef struct node Node;
struct node
{
    int data;
    Node *previous, *next;
};

// Create A Node...
Node *create_node(int item, Node *previous, Node *next)
{
    Node *new_node = (Node *)malloc(sizeof(Node));
    if(new_node == NULL) exit(1);
    new_node->data = item;
    new_node->previous = previous;
    new_node->next = next;
    return new_node;
}

// Remove A Node...
Node *remove_node(Node *head, Node *node)
{
    if(node == head)
    {
        head = node->next;
        free(node);
        return head;
    }
    node->previous->next = node->next;
    free(node);
    return head;
}

// Insert A Node At The Beginning...
Node *prepend(Node *head, int item)
{
    Node *new_node = create_node(item, NULL, head);
    return new_node;
}

// Insert A Node...
void *insert(Node *node, int item)
{
    Node *new_node = create_node(item, node, node->next);
    node->next = new_node;
}

// Insert A Node At The End...
Node *append(Node *head, int item)
{
    if(head == NULL) return create_node(item, NULL, NULL); 
    Node *current_node = head;
    while(current_node->next != NULL)
        current_node = current_node->next;
    Node *new_node = create_node(item, current_node, NULL);
    current_node->next = new_node;
    return head;
}

// Find A Node
Node *find_node(Node *head, int item)
{
    Node *temp = head;
    while(temp != NULL)
    {
        if(temp->data == item) return temp;
        temp = temp->next;
    }
    return NULL;
}

// Print Linked_List...
void print_linked_list(Node *head)
{
    Node *current_node = head;
    printf("[%d", current_node->data);
    current_node = current_node->next;
    while(current_node != NULL)
    {
        printf(", %d", current_node->data);
        current_node = current_node->next;
    }
    printf("]\n");
}


//-------------------------------------------------------------------------------//


int main()
{
    Node *linked_list = create_node(0, NULL, NULL);

    append(linked_list, 1);
    append(linked_list, 2);
    append(linked_list, 3);
    append(linked_list, 4);
    append(linked_list, 5);

    print_linked_list(linked_list);
    return 0;
}

File: Data_Structures_And_Algorithms/Python/Data_Structure/doubly_linked_list.py
# Doubly_Linked_List Data_Structure

class Node:
    def __init__(self, data, previous=None, next=None):
        self.data = data
        self.previous = previous
        self.next = next
    
    # Remove A Node...
    def remove_node(self, node):
        if node == self:
            self = self.next
            return self
        node.previous.next = node.next
        return self

    # Insert A Node At The Beginning...
    def prepend(self, item):
        return Node(item, None, self)
    
    # Insert A Node...
    def insert(self, node, item):
        node.next = Node(item, node, node.next)

    # Insert A Node At The End...
    def append(self, item):
        new_node = Node(item)
        if self == None: return new_node
        current_node = self
        while current_node.next != None:
            current_node = current_node.next
        current_node.next = new_node
        return self

    # Find A Node
    def find_node(self, item):
        temp = self
        while temp != None:
            if temp.data == item: return temp
            temp = temp.next
        return None
    
    # Print Linked_List...
    def print_linked_list(self):
        current_node = self
        print(f'[{current_node.data}', end='')
        current_node = current_node.next
        while current_node != None:
            print(f', {current_node.data}', end='')
            current_node = current_node.next
        print(']')


#-------------------------------------------------------------------------------#


head = Node(0)
head.append(1)
head.append(2)
head.append(3)
head.append(4)
head.append(5)
head.print_linked_list()